File: scala/raindrops/src/main/scala/Raindrops.scala
object Raindrops {
  val pliaong = Map(3 -> "Pling", 5 -> "Plang", 7 -> "Plong")

  def convert(nr:Long) = {
    pliaong.foldRight(Seq.empty[String]){
      case ((k, v), acc) if nr%k == 0 => v +: acc
      case (_, acc) => acc
    } match {
      case xs if xs.isEmpty => nr.toString
      case xs => xs.mkString
    }
  }

  def apply() = Raindrops
}

File: haskell/raindrops/src/Raindrops.hs
module Raindrops (convert) where

import qualified Data.Map as M

pliaong :: M.Map Int String
pliaong = M.fromList [(3, "Pling"), (5, "Plang"), (7, "Plong")]

convert :: Int -> String
convert n =
  case (pliaong M.!) <$> factors n of
    [] -> show n
    xs -> xs >>= id
  where
    factors x = [ i | i <- M.keys pliaong, x `mod` i == 0]
